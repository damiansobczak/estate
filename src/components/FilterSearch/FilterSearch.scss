@import "../../assets/styles/utils/variables";
@import "../../assets/styles/utils/mixins";

.filterSearch {
  &__form {
    display: flex;
    background-color: #fff;
    border-radius: 4px;
    overflow: hidden;
    box-shadow: 0 20px 40px rgba(0, 0, 0, 0.1);
    font-family: poppins, sans-serif;
    max-width: 1100px;
    margin: 0 auto;
  }

  &__wrapper {
    display: flex;
    flex: 1;
  }

  &__group {
    display: flex;
    align-items: center;
    padding: 0 30px;
    margin: 10px 0;
    flex: 1;

    @include breakpoint-down(sm) {
      padding: 0 15px;
    }

    &:nth-child(2) {
      border-left: 1px solid rgba($color-neutral-300, 0.2);
      border-right: 1px solid rgba($color-neutral-300, 0.2);
    }

    &--price {
      @include breakpoint-down(lg) {
        display: none;
      }
    }

    &--category {
      @include breakpoint-down(lg) {
        display: none;
      }
    }
  }

  &__label {
    font-weight: 600;
    margin-right: 15px;
  }

  &__input {
    border: 0;
    outline: 0;
    font-size: $font-200;
    color: $color-text-secondary;
    display: flex;
    flex: 1;

    &::placeholder {
      color: $color-text-secondary;
    }
  }

  &__btn {
    background-color: $color-accent;
    color: #fff;
    font-size: $font-200;
    padding: 18px 40px;
    margin-left: auto;
    transition: background-color 0.2s ease-in-out;

    &:hover {
      background-color: darken($color-accent, 7%);
    }
  }
}
