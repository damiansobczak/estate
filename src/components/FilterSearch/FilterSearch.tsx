import React, { FormEvent } from "react";
import "./FilterSearch.scss";
import { getSearches } from "../../API";
import FilterResult from "../FilterResult/FilterResult";

class FilterSearch extends React.Component {
  node = React.createRef<any>();
  form = React.createRef<HTMLFormElement>();

  state = {
    isShowing: false,
    result: null,
    text: '',
    price: '',
    category: ''
  }

  search = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const text = this.state.text;
    const price = this.state.price && Number(this.state.price) !== 0 ? Number(this.state.price.replace(/zł|,| /g, '')) : 999999999;
    const category = this.state.category;
    const data = await getSearches(text, price, category);
    this.setState({
      isShowing: !this.state.isShowing,
      result: data
    });
  }

  clickOutside = (ev: any) => {
    if (this.state.isShowing && !this.node.current.contains(ev.target)) {
      this.setState({
        isShowing: false
      });
    }
  }

  updateField = (field: any) => (e: any) => {
    this.setState({
      [field]: e.target.value
    });
  }

  componentWillMount() {
    document.addEventListener('mousedown', this.clickOutside, false);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.clickOutside, false);
  }

  render() {
    return (
      <div className="filterSearch container" ref={this.node}>
        <form className="filterSearch__form" ref={this.form} onSubmit={this.search}>
          <div className="filterSearch__wrapper">
            <div className="filterSearch__group">
              <label htmlFor="filterSearch__name" className="filterSearch__label">
                Szukaj
              </label>
              <input id="filterSearch__name" type="text" className="filterSearch__input" value={this.state.text} placeholder="Napisz nam czego szukasz?" onChange={this.updateField('text')} />
            </div>
            <div className="filterSearch__group filterSearch__group--price">
              <label htmlFor="filterSearch__price" className="filterSearch__label">
                Cena
              </label>
              <input type="text" id="filterSearch__price" className="filterSearch__input" value={this.state.price} placeholder="np. 200,000zł" onChange={this.updateField('price')} />
            </div>
            <div className="filterSearch__group filterSearch__group--category">
              <label htmlFor="filterSearch__category" className="filterSearch__label">
                Kategoria
              </label>
              <input id="filterSearch__category" type="text" className="filterSearch__input" value={this.state.category} placeholder="Kupno lub wynajem" onChange={this.updateField('category')} />
            </div>
          </div>
          <button className="filterSearch__btn">Wyszukaj</button>
        </form>
        <FilterResult isShowing={this.state.isShowing} data={this.state.result} />
      </div>
    );
  }
}

export default FilterSearch;
