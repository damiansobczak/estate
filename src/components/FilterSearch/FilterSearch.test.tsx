import React from "react";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import FilterSearch from "./FilterSearch";

describe("<FilterSearch />", () => {
    it("renders component correctly", () => {
        const tree = renderer.create(<FilterSearch />).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it("allow to search by name", () => {
        const wrapper = shallow(<FilterSearch />);
        let input = wrapper.find("input#filterSearch__name");
        input.simulate('change', { target: { value: 'Osiedle' } });
        expect(wrapper.find("input#filterSearch__name").props().value).toEqual("Osiedle");
    });

    it("allow to search by price", () => {
        const wrapper = shallow(<FilterSearch />);
        let input = wrapper.find("input#filterSearch__price");
        input.simulate('change', { target: { value: '200,000' } });
        expect(wrapper.find("input#filterSearch__price").props().value).toEqual("200,000");
    });

    it("allow to search by category", () => {
        const wrapper = shallow(<FilterSearch />);
        let input = wrapper.find("input#filterSearch__category");
        input.simulate('change', { target: { value: 'Wynajem' } });
        expect(wrapper.find("input#filterSearch__category").props().value).toEqual("Wynajem");
    });

    it("updates state on input change", () => {
        const wrapper = mount(<FilterSearch />);
        let name = wrapper.find("input#filterSearch__name");
        let price = wrapper.find("input#filterSearch__price");
        let category = wrapper.find("input#filterSearch__category");

        name.simulate('change', { target: { value: 'Osiedle' } });
        price.simulate('change', { target: { value: '999' } });
        category.simulate('change', { target: { value: 'Wynajem' } });

        expect(wrapper.state('text')).toEqual('Osiedle');
        expect(wrapper.state('price')).toEqual('999');
        expect(wrapper.state('category')).toEqual('Wynajem');
    });
});
