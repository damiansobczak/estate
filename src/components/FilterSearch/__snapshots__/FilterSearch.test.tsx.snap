// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterSearch /> renders component correctly 1`] = `
<div
  className="filterSearch container"
>
  <form
    className="filterSearch__form"
    onSubmit={[Function]}
  >
    <div
      className="filterSearch__wrapper"
    >
      <div
        className="filterSearch__group"
      >
        <label
          className="filterSearch__label"
          htmlFor="filterSearch__name"
        >
          Szukaj
        </label>
        <input
          className="filterSearch__input"
          id="filterSearch__name"
          onChange={[Function]}
          placeholder="Napisz nam czego szukasz?"
          type="text"
          value=""
        />
      </div>
      <div
        className="filterSearch__group filterSearch__group--price"
      >
        <label
          className="filterSearch__label"
          htmlFor="filterSearch__price"
        >
          Cena
        </label>
        <input
          className="filterSearch__input"
          id="filterSearch__price"
          onChange={[Function]}
          placeholder="np. 200,000zÅ‚"
          type="text"
          value=""
        />
      </div>
      <div
        className="filterSearch__group filterSearch__group--category"
      >
        <label
          className="filterSearch__label"
          htmlFor="filterSearch__category"
        >
          Kategoria
        </label>
        <input
          className="filterSearch__input"
          id="filterSearch__category"
          onChange={[Function]}
          placeholder="Kupno lub wynajem"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="filterSearch__btn"
    >
      Wyszukaj
    </button>
  </form>
  <div
    className="filterResult  "
  >
    <div
      className="filterResult__title"
    >
      Brak ofert
    </div>
  </div>
</div>
`;
